cpp/src/arrow/csv/writer.cc

```diff
@@ -0,0 +1,398 @@
+// This class handles escaping assuming that all strings will be quoted
+// and that the only character within the string that needs to escaped is
+// a quote character (") and escaping is done my adding another quote.
+class QuotedColumnPopulator : public ColumnPopulator {
+ public:
+  QuotedColumnPopulator(MemoryPool* pool, char end_char)
+      : ColumnPopulator(pool, end_char) {}
+
+  Status UpdateRowLengths(int32_t* row_lengths) override {
+    const StringArray& input = *casted_array_;
+    extra_chars_count_.resize(input.length());
+    auto extra_chars = extra_chars_count_.begin();
+    VisitArrayDataInline<StringType>(
+        *input.data(),
+        [&](arrow::util::string_view s) {
+          int64_t escaped_count = CountEscapes(s);
+          // TODO: Maybe use 64 bit row lengths or safe cast?
+          *extra_chars = static_cast<int>(escaped_count) + kQuoteCount;
+          extra_chars++;
+        },
+        [&]() {
+          *extra_chars = 0;
+          extra_chars++;
+        });
+
+    for (int x = 0; x < input.length(); x++) {
+      row_lengths[x] += extra_chars_count_[x] + input.value_length(x);
+    }
+    return Status::OK();
+  }
+
+  void PopulateColumns(char** rows) const override {
+    const int32_t* extra_chars = extra_chars_count_.data();
+    VisitArrayDataInline<StringType>(
+        *(casted_array_->data()),
+        [&](arrow::util::string_view s) {
+          int64_t next_column_offset = *extra_chars + s.length() + /*end_char*/ 1;
+          **rows = '"';
+          if (*extra_chars == kQuoteCount) {
+            memcpy((*rows + 1), s.data(), s.length());
+          } else {
+            Escape(s, (*rows + 1));
+          }
+          *(*rows + next_column_offset - 2) = '"';
+          *(*rows + next_column_offset - 1) = end_char_;
+          *rows += next_column_offset;
+          extra_chars++;
+          rows++;
+        },
+        [&]() {
+          // Nulls are empty (unquoted) to distinguish with empty string.
+          **rows = end_char_;
+          *rows += 1;
+          rows++;
+          extra_chars++;
+        });
+  }
+
+ private:
+  std::vector<int32_t, std::allocator<int32_t>> extra_chars_count_;
+};
+
+struct PopulatorFactory {
+  template <typename TypeClass>
+  enable_if_t<is_base_binary_type<TypeClass>::value ||
+                  std::is_same<FixedSizeBinaryType, TypeClass>::value,
+              Status>
+  Visit(const TypeClass& type) {
+    populator = new QuotedColumnPopulator(pool, end_char);
+    return Status::OK();
+  }
+
+  template <typename TypeClass>
+  enable_if_dictionary<TypeClass, Status> Visit(const TypeClass& type) {
+    return VisitTypeInline(*type.value_type(), this);
+  }
+
+  template <typename TypeClass>
+  enable_if_t<is_nested_type<TypeClass>::value || is_extension_type<TypeClass>::value,
+              Status>
+  Visit(const TypeClass& type) {
+    return Status::Invalid("Nested and extension types not supported");
```

Comment:
You can probably use a fallback `Visit` method instead of spelling out the type condition above.